src/Utilities/TaskItem.cs

```diff
@@ -298,9 +298,23 @@ public void CopyMetadataTo(ITaskItem destinationItem)
             if (_metadata != null)
             {
                 // Avoid a copy if we can
-                if (destinationItem is TaskItem destinationAsTaskItem && destinationAsTaskItem.Metadata == null)
+                if (destinationItem is TaskItem destinationAsTaskItem)
                 {
-                    destinationAsTaskItem.Metadata = _metadata.Clone(); // Copy on write!
+                    CopyOnWriteDictionary<string> copiedMetadata;
+                    if (destinationAsTaskItem.Metadata == null || destinationAsTaskItem.Metadata.Count < _metadata.Count)
+                    {
+                        copiedMetadata = _metadata.Clone(); // Copy on write!
+                        if (destinationAsTaskItem.Metadata != null)
+                        {
+                            copiedMetadata.SetItems(destinationAsTaskItem.Metadata);
+                        }
+                    }
+                    else
+                    {
+                        copiedMetadata = destinationAsTaskItem.Metadata.Clone();
+                        copiedMetadata.SetItems(Metadata.Where(entry => !destinationAsTaskItem.Metadata.ContainsKey(entry.Key)));
```

Comment:
I was unsure between using Where or setting them one-by-one in this case. I did some benchmarking, and this is what I got:
![image](https://user-images.githubusercontent.com/12969783/101220256-37f8fe00-363a-11eb-8eb1-cb279d03f3c1.png)
Note that for each test case, I added `NumStrings / 10` items to an ImmutableDictionary with `NumStrings` items in it previously.

You can see that using Where had fewer allocations for all three sizes. (Base is creating but not combining the dictionaries, so Where actually allocated very little.) It lost on time, though, unless the dataset was large, and I don't have data as to whether that's expected sometimes, always, or never. Thoughts?

Reply:
Can you instead do `copiedMetadata = Metadata.Clone().SetItems(destinationAsTaskItem.Metadata)`?

Reply:
Yes, and honestly, there are probably a lot of cases in which that would be better than what I currently have. What I'm trying to optimize for is adding metadata from a small dictionary to a large dictionary, and in that case, it would be better to add a few items individually rather than using SetItems to set a lot of items. I can run a test for that, though, and see if it's optimized nicely, in which case I'll switch to that.

Reply:
It looks like doing it that way was 50-75% slower if the adding dictionary had 10x as many items. Next, trying it with a smaller ratio...

Reply:
I tried comparing adding the larger dictionary to the smaller dictionary for ratios of 1, 2, 5, and 8, and it looks like setting with `Where` was faster for all but 1x size. I currently think that's the best option.
![image](https://user-images.githubusercontent.com/12969783/101226494-7fd35180-3649-11eb-991d-f0f2b9dfcab1.png)